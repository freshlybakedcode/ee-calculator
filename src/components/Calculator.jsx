import React, { useState } from 'react';
import styled from 'styled-components';

import { TopSection } from './TopSection';
import { CalculatorScreen } from './CalculatorScreen';
import { Keypad } from './Keypad';

import { convertOperator } from '../utils/convertOperator';
import { isMathmaticalOperator } from '../utils/isMathmaticalOperator';
import { evaluateSum } from '../utils/evaluateSum';
import { clearEntry } from '../utils/clearEntry';

const Chassis = styled.div`
  box-shadow: 3px 2px 12px 0px #333;
  overflow: hidden;
  background: linear-gradient(to right, #6d6678 0%, #352e40 5%);
  border-radius: 20px;
  display: flex;
  flex-direction: column;
  flex: 1;
  max-width: 300px;
`;

export const Calculator = () => {
  const [currentSum, setCurrentSum] = useState('');
  const [displayedDigits, setDisplayedDigits] = useState('0');

  const handleEquals = () => {
    if (
      currentSum !== '' &&
      !isMathmaticalOperator(currentSum.charAt(currentSum.length - 1))
    ) {
      const returnedAnswer = evaluateSum(currentSum);
      setCurrentSum(returnedAnswer);
      setDisplayedDigits(returnedAnswer);
    }
  };

  const handleClear = (clear) => {
    if (clear === 'C') {
      setCurrentSum('');
      setDisplayedDigits('0');
    } else {
      setCurrentSum(clearEntry(currentSum));
      setDisplayedDigits('0');
    }
  };

  const handleDecimal = () => {
    if (displayedDigits.indexOf('.') === -1) {
      setDisplayedDigits(displayedDigits + '.');
      setCurrentSum(currentSum + '.');
    }
  };

  const handleOperator = (operator) => {
    if (operator === '=') {
      // if the equals is pressed then evaluate sum
      handleEquals();
    } else if (operator === '.') {
      handleDecimal();
    } else if (operator === 'C' || operator === 'CE') {
      handleClear(operator);
    } else if (
      isMathmaticalOperator(currentSum.charAt(currentSum.length - 1))
    ) {
      // if the most recent entry in the currentSum is an operator then change the exisiting operator to the newly pressed one
      setCurrentSum(currentSum.slice(0, currentSum.length - 1) + operator);
    } else if (
      isMathmaticalOperator(currentSum.slice(0, currentSum.length - 1))
    ) {
      // if this is the second operator used in the buffer, then evaluate and then add the next operator
      handleEquals();
      setCurrentSum(currentSum + convertOperator(operator));
    } else {
      if (currentSum !== '') {
        setCurrentSum(currentSum + convertOperator(operator));
      }
    }
  };

  const handleNumericalInput = (newDigit) => {
    // if the only entry is a zero then remove the zero
    if (displayedDigits === '0' && currentSum === '') {
      setDisplayedDigits(newDigit);
      setCurrentSum(newDigit);
    } else if (
      isMathmaticalOperator(
        convertOperator(currentSum.charAt(currentSum.length - 1))
      )
    ) {
      // clear display when proceeding after an operator
      setDisplayedDigits(newDigit);
      setCurrentSum(currentSum + newDigit);
    } else {
      setDisplayedDigits(displayedDigits + newDigit);
      setCurrentSum(currentSum + newDigit);
    }
  };

  const handleKeypadPress = (e) => {
    const buttonPressed = e.target.innerHTML;
    const operator = e.target.dataset.operator;
    operator
      ? handleOperator(buttonPressed)
      : handleNumericalInput(buttonPressed);
  };

  return (
    <Chassis>
      <TopSection />
      <CalculatorScreen display={displayedDigits} />
      <Keypad handleKeypadPress={handleKeypadPress} />
    </Chassis>
  );
};
