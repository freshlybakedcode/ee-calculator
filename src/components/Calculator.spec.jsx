import { fireEvent, render } from '@testing-library/react';
import { Calculator } from './Calculator';

describe('Calculator', () => {
  let buttonOne,
    buttonTwo,
    buttonAdd,
    buttonSubtract,
    buttonEquals,
    buttonC,
    buttonCE,
    buttonDecimal,
    output;

  beforeEach(() => {
    const { getByRole, getByTestId } = render(<Calculator />);
    buttonOne = getByRole('button', {
      name: /1/,
    });
    buttonTwo = getByRole('button', {
      name: /2/,
    });
    buttonAdd = getByRole('button', {
      name: /\+/,
    });
    buttonSubtract = getByRole('button', {
      name: /-/,
    });
    buttonEquals = getByRole('button', {
      name: /=/,
    });
    buttonC = getByRole('button', {
      name: /^C$/,
    });
    buttonCE = getByRole('button', {
      name: /CE/,
    });
    buttonDecimal = getByRole('button', {
      name: /\./,
    });
    output = getByTestId('calculator-screen');
  });

  const onePlusOne = () => {
    fireEvent.click(buttonOne);
    fireEvent.click(buttonAdd);
    fireEvent.click(buttonOne);
  };

  it('should render as expected', () => {
    const { container } = render(<Calculator />);
    expect(container).toMatchSnapshot();
  });

  describe('Operation', () => {
    it('should be able to add two numbers together and display the output', () => {
      onePlusOne();
      fireEvent.click(buttonEquals);
      expect(output).toHaveTextContent('2');
    });
    it('should change the last operator if the most recent entry in the current sum is an operator but a new operator is chosen by the user', () => {
      fireEvent.click(buttonTwo);
      fireEvent.click(buttonAdd);
      fireEvent.click(buttonSubtract);
      fireEvent.click(buttonOne);
      fireEvent.click(buttonEquals);
      expect(output).toHaveTextContent('1');
    });
    it('should perform evaluation on the sum if there is already an operator in the current sum, the last character in the current sum is not an operator, and the user chooses an additional operator', () => {
      fireEvent.click(buttonTwo);
      fireEvent.click(buttonAdd);
      fireEvent.click(buttonOne);
      fireEvent.click(buttonSubtract);
      expect(output).toHaveTextContent('3');
    });
  });

  describe('Clear functions', () => {
    it('should reset to 0 when the clear button is pressed', () => {
      onePlusOne();
      fireEvent.click(buttonC);
      expect(output).toHaveTextContent('0');
    });

    it('should clear only the current entry when the Clear Entry button is pressed', () => {
      onePlusOne();
      fireEvent.click(buttonCE);
      expect(output).toHaveTextContent('0');
      fireEvent.click(buttonTwo);
      fireEvent.click(buttonEquals);
      expect(output).toHaveTextContent('3');
    });
  });

  describe('Decimal functions', () => {
    it('should be able to add a decimal', () => {
      fireEvent.click(buttonOne);
      fireEvent.click(buttonDecimal);
      fireEvent.click(buttonTwo);
      expect(output).toHaveTextContent('1.2');
    });
    it('should not be able to add multiple decimal points to the same number', () => {
      fireEvent.click(buttonOne);
      fireEvent.click(buttonDecimal);
      fireEvent.click(buttonDecimal);
      fireEvent.click(buttonTwo);
      fireEvent.click(buttonDecimal);
      expect(output).toHaveTextContent('1.2');
    });
  });

  describe('Inputs', () => {
    it('should remove the zero when adding a number if the only entry is a zero', () => {
      expect(output).toHaveTextContent('0');
      fireEvent.click(buttonOne);
      expect(output).toHaveTextContent('1');
    });
    it('should clear the display after an operator is selected and a new digit is entered', () => {
      fireEvent.click(buttonOne);
      fireEvent.click(buttonAdd);
      fireEvent.click(buttonTwo);
      expect(output).toHaveTextContent('2');
    });
  });
});
