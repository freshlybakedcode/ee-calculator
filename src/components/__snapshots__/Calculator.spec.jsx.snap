// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Calculator should render as expected 1`] = `
<div>
  <div
    class="sc-jrsJWt bGtHQU"
  >
    <div
      class="sc-bdnxRM cxLqHU"
    >
      <div
        class="sc-gtsrHT lkAkzf"
      >
        <div
          class="sc-dlnjwi hhIhYm"
        />
        <div
          class="sc-dlnjwi hhIhYm"
        />
        <div
          class="sc-dlnjwi hhIhYm"
        />
        <div
          class="sc-dlnjwi hhIhYm"
        />
      </div>
    </div>
    <div
      class="sc-hKFxyN hFgbwT"
    >
      <div
        class="sc-eCApnc bzHlTF"
      >
        <img
          alt="[=]"
          class="sc-jSFjdj iURHSZ"
          src="logo.svg"
        />
      </div>
      <span
        class="sc-gKAaRy jrThlc"
        data-testid="calculator-screen"
      >
        0
      </span>
    </div>
    <div
      class="sc-iCoGMd bvKOAP"
    >
      <button
        class="sc-fujyAs gVNkNN"
        data-operator="clear"
      >
        C
      </button>
      <button
        class="sc-fujyAs gVNkNN"
        data-operator="clearEntry"
      >
        CE
      </button>
      <div
        class="sc-pNWdM ghGQzl"
      />
      <button
        class="sc-fujyAs ySqhB"
      >
        7
      </button>
      <button
        class="sc-fujyAs ySqhB"
      >
        8
      </button>
      <button
        class="sc-fujyAs ySqhB"
      >
        9
      </button>
      <button
        class="sc-fujyAs ySqhB"
        data-operator="divide"
      >
        รท
      </button>
      <button
        class="sc-fujyAs ySqhB"
      >
        4
      </button>
      <button
        class="sc-fujyAs ySqhB"
      >
        5
      </button>
      <button
        class="sc-fujyAs ySqhB"
      >
        6
      </button>
      <button
        class="sc-fujyAs ySqhB"
        data-operator="multiply"
      >
        ร
      </button>
      <button
        class="sc-fujyAs ySqhB"
      >
        1
      </button>
      <button
        class="sc-fujyAs ySqhB"
      >
        2
      </button>
      <button
        class="sc-fujyAs ySqhB"
      >
        3
      </button>
      <button
        class="sc-fujyAs ySqhB"
        data-operator="subtract"
      >
        -
      </button>
      <button
        class="sc-fujyAs ySqhB"
      >
        0
      </button>
      <button
        class="sc-fujyAs ySqhB"
        data-operator="decimal"
      >
        .
      </button>
      <button
        class="sc-fujyAs ySqhB"
        data-operator="equals"
      >
        =
      </button>
      <button
        class="sc-fujyAs ySqhB"
        data-operator="add"
      >
        +
      </button>
    </div>
  </div>
</div>
`;
